packages/uma-auth-client/src/hooks/useOAuth.tsx

```diff
@@ -0,0 +1,100 @@
+import {auth, nwc, webln} from '@getalby/sdk';
+import { useEffect, useState } from "react";
+import { getUmaDomain } from "src/utils/getUmaDomain";
+import { useUser } from "./useUser";
+import { create } from 'zustand';
+import { persist } from "zustand/middleware";
+import { getNwcUmaDomain } from 'src/utils/getNwcUmaDomain';
+
+type UmaAuthToken = {
+  access_token: string,
+  refresh_token: string,
+  expires_at: number,
+  nwc_connection_uri: string;
+  commands: string[];
+  budget: string;
+  nwc_expires_at: string;
+};
+
+interface TokenState {
+  accessToken: string;
+  refreshToken: string;
+  expiresAt: number;
+}
+
+interface OAuthState {
+  codeVerifier?: string;
+  token?: TokenState;
+  identityNpub?: string;
+  identityRelayUrl?: string;
+  redirectUri?: string;
+  nwcConnectionUri?: string;
+  setToken: (token: TokenState) => void;
+  performOAuth: (uma: string) => Promise<void>;
+}
+
+export const useOAuth = create<OAuthState>()(
+  persist(
+    (set, get) => ({
+      setToken: (token) => set({
+        token,
+      }),
+      performOAuth: async (uma) => {
+        const state = get();
+        const result = await performOAuth(state, uma);
+        set(result);
+      },
+    }),
+    {
+      name: "uma-connect",
+      partialize: (state) => ({
+        nwcConnectionUri: state.nwcConnectionUri,
+      })
+    }
+  )
+);
+
+const performOAuth = async (state: OAuthState, uma: string) => {
+  const { identityNpub, identityRelayUrl, redirectUri, codeVerifier, token, nwcConnectionUri } = state;
+
+  // If we have a connection URI and a token that hasn't expired, we don't need to do anything
+  if (nwcConnectionUri && token && (!token.expiresAt || token.expiresAt < Date.now())) {
+    return state;
+  }
+
+  const authClient = new auth.OAuth2User({
+    client_id: `${identityNpub} ${identityRelayUrl}`,
+    callback: redirectUri!, // TODO: get this from client app
+    scopes: [],
+    user_agent: "uma-connect",
+    request_options: {
+      base_url: getNwcUmaDomain(uma),
```

Comment:
We may need to retrieve a url from the discovery config here, but I know you still have some research to do on how the alby sdk deals with these urls.

Btw, for the auth request, my reference in bitcoin-connect is [right here](https://github.com/jklein24/bitcoin-connect/blob/c3de26d3ae2799d5b5772811b4b39f3889009319/src/components/pages/bc-uma.ts#L78). When the redirect comes through, that's when it's [handled by the connector](https://github.com/jklein24/bitcoin-connect/blob/feat/crosscurrency/src/connectors/UMANWCConnector.ts).

Reply:
Oh so if i'm understanding right, the connector class is just used after the redirect and for retrieving the tokens, while the bc-uma code is actually the start of the oauth handshake

Reply:
Yep, correct.